docs/running-on-kubernetes.md

```diff
@@ -208,8 +208,30 @@ If your application's dependencies are all hosted in remote locations like HDFS
 by their appropriate remote URIs. Also, application dependencies can be pre-mounted into custom-built Docker images.
 Those dependencies can be added to the classpath by referencing them with `local://` URIs and/or setting the
 `SPARK_EXTRA_CLASSPATH` environment variable in your Dockerfiles. The `local://` scheme is also required when referring to
-dependencies in custom-built Docker images in `spark-submit`. Note that using application dependencies from the submission
-client's local file system is currently not yet supported.
+dependencies in custom-built Docker images in `spark-submit`. We support dependencies from the submission
+client's local file system using the `file://` scheme or without a scheme (using a full path), where the destination should be a Hadoop compatible filesystem.
+A typical example of this using S3 is via passing the following options:
+
+```
+...
+--packages com.amazonaws:aws-java-sdk:1.7.4,org.apache.hadoop:hadoop-aws:2.7.6
+--conf spark.kubernetes.file.upload.path=s3a://<s3-bucket>/path
+--conf spark.hadoop.fs.s3a.access.key=...
+--conf spark.hadoop.fs.s3a.impl=org.apache.hadoop.fs.s3a.S3AFileSystem
+--conf spark.hadoop.fs.s3a.fast.upload=true
+--conf spark.hadoop.fs.s3a.secret.key=....
+--conf spark.driver.extraJavaOptions=-Divy.cache.dir=/tmp -Divy.home=/tmp
+file:///full/path/to/app.jar
+```
+The app jar file will be uploaded to the S3 and then when the driver is launched it will be downloaded
+to the driver pod and will be added to its classpath.
+
+The client scheme is supported for the application jar, and dependencies specified by properties `spark.jars` and `spark.files`.
+
+Important: all client-side dependencies will be uploaded to the given path with a flat directory structure so
```

Comment:
One thing that I'm missing after reading this again is how is cleanup handled? What about multiple applications being started concurrently?

Seems like you'll end up with a directory full or jars from different applications, and two of them decide to upload "app.jar", one of them might fail because it's using the wrong jar or a partially written one (because the second submission is still happening, e.g.).

Wouldn't it be better to use a submission-specific path that Spark creates, and have the driver make a best effort to clean it up after it downloads the dependencies? Similar to what Spark does on YARN.

Reply:
@vanzin Users can specify the path for uploading files. They should pick a different dir per app if they are going to use the same filenames for the files uploaded. There is no cleanup after the job finished, user should cleanup the app dir in the dfs.

Reply:
That's a pretty horrible user experience. Normally there is going to be a configuration file that has this path hardcoded in it along with a lot of other configuration that doesn't change.

Creating a unique path is super easy, even if you don't bother to clean it up later. Adding the clean up just avoids yet another thing the user needs to remember to do. After all, storage in the cloud is not free...

Reply:
It is a pretty horrible strategy/attitude for a developer not to separate where you put your apps per job.

Reply:
I don't think it's horrible either way, doesn't need to be framed this way. If there's no cost to buying the user some extra isolation, that seems slightly better. If there is, let's discuss the downside

Reply:
>  Each time I launch my job I pass a path thats all.

Exactly, that's what I'm saying is a bad user experience. Why do you need extra work to make separate jobs don't mess with each other?

Look at what the YARN backend does. There's a root (the equivalent of a temp directory in Sean's example), and each app creates it own separate directory inside it. So you don't need to change your app invocation every time you want to run a new app. It's trivial code to add in Spark that saves people a lot of headaches.